File: python/dfsAndBfs.py
# DFS: explores as deep as possible along a branch before backtracking 
def dfs(graph, start):
  visited = set()
  stack = [start]

  while stack:
    vertex = stack.pop()
    if vertex not in visited:
      visited.add(vertex)
      print(vertex)
      for neighbor in graph[vertex]: 
        stack.append(neighbor)

# example usage
graph - {
  'A': ['B', 'C'],
  'B': ['D', 'E'],
  'C: ['F'],
  'D': [],
  'E': ['F'],
  'F' : []
}

dfs(graph, 'A')

# BFS: explores all neighbors at present depth before moving on to next depth level 
from collections import deque
def bfs(graph, start):
  visited = set()
  queue = deque([start])

  while queue:
    vertex = queue.popleft()
    if vertex not in visited:
      visited.add(vertex)
      print(vertex)
      for neighbor in graph[vertex]:
        queue.append(neighbor)

# example usage
bfs(graph, 'A')

File: typescript/dfsAndBfs.ts
function dfs<T>(graph: Map<T, T[]>, start: T, visited: Set<T> = new Set()): void {
  visited.add(start);
  console.log(start);

  for(const neighbor of graph.get(start)!) {
    if(!visited.has(neighbor)) {
      dfs(graph, neighbor, visited);
    }
  }
}

function bfs<T>(graph: Map<T, T[]>, start: T): void {
  const queue: T[] = [start];
  const visited = new Set<T>();
  visited.add(start);

  while(queue.length > 0) {
    const current = queue.shift()!;
    console.log(current);

    for(const neighbor of graph.get(current)!) {
      if(!visited.has(neighbor)) {
        visited.add(neighbor);
        queue.push(neighbor);
      }
    }
  }
}


// example usage
const graph = new Map<string, string[]>();
graph.set('A', ['B', 'C']);
graph.set('B', ['D', 'E']);
graph.set('C', ['F']);
graph.set('D', ['G']);
graph.set('E', ['H']);
graph.set('F', ['I']);
graph.set('G', ['J']);
graph.set('H', []);
graph.set('I', []);
graph.set('J', []);

console.log("dfs traversal: ");
dfs(graph, 'A');

console.log("\nBFS traversal:");
bfs(graph, 'A');

